unblob/handlers/archive/qnap/qnap_nas.py

```diff
@@ -0,0 +1,203 @@
+from structlog import get_logger
+
+from unblob.file_utils import File, iterate_file, stream_scan
+from unblob.models import (
+    Endian,
+    Extractor,
+    HexString,
+    StructHandler,
+    StructParser,
+    ValidChunk,
+)
+
+logger = get_logger()
+
+FOOTER_LEN = 74
+SECRET = "QNAPNASVERSION"
+
+C_DEFINITIONS = """
+    typedef struct qnap_header {
+        char    magic[6];
+        uint32  encrypted_len;
+        char    device_id[16];
+        char    file_version[16];
+        char    firmware_date[16];
+        char    revision[16];
+    } qnap_header_t;
+"""
+FOOTER_PATTERN = [
+    HexString("69 63 70 6e 61 73"),
+]
+
+
+@attr.define
+class QTSSearchContext:
+    start_offset: int
+    file: File
+    end_block_offset: int
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.encrypted_len < len(header):
+        return False
+    try:
+        header.device_id.decode("utf-8")
+        header.file_version.decode("utf-8")
+        header.firmware_date.decode("utf-8")
+        header.revision.decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+def _hyperscan_match(
+    context: QTSSearchContext, pattern_id: int, offset: int, end: int
+) -> Scan:
+    if offset < context.start_offset:
+        return Scan.Continue
+    struct_parser = StructParser(C_DEFINITIONS)
+    context.file.seek(offset, io.SEEK_SET)
+    header = struct_parser.parse("qnap_header_t", context.file, Endian.LITTLE)
+    logger.debug("qnap_header_t", header=header)
+
+    if is_valid_header(header):
+        context.end_block_offset = context.file.tell()
+        return Scan.Terminate
+    return Scan.Continue
+
+
+def build_stream_end_scan_db(pattern_list):
+    return StreamDatabase(
+        *(Pattern(p.as_regex(), Flag.SOM_LEFTMOST, Flag.DOTALL) for p in pattern_list)
+    )
+
+
+hyperscan_stream_end_magic_db = build_stream_end_scan_db(FOOTER_PATTERN)
+
+
+class QnapExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        outpath = outdir.joinpath(f"{inpath.name}.decrypted")
+        with File.from_path(inpath) as file:
+            file.seek(-FOOTER_LEN, io.SEEK_END)
+            header = self._struct_parser.parse("qnap_header_t", file, Endian.LITTLE)
+            end_offset = file.tell()
+            filesize = end_offset
+            cryptor = Cryptor(SECRET + header.file_version.decode("utf-8")[0])
+            with outpath.open("wb") as outfile:
+                for chunk in iterate_file(file, 0, header.encrypted_len, 1024):
+                    outfile.write(cryptor.decrypt_chunk(chunk))
+                for chunk in iterate_file(
+                    file,
+                    header.encrypted_len,
+                    filesize - FOOTER_LEN - header.encrypted_len,
+                    1024,
+                ):
+                    outfile.write(chunk)
+
+
+class QnapHandler(StructHandler):
```

Comment:
You're no longer parsing based on C definitions here.

```suggestion
class QnapHandler(Handler):
```